mlir/lib/Gradient/Transforms/vector_lowering.cpp

```diff
@@ -0,0 +1,231 @@
+            {dataField.getType(), sizeField.getType(), capacityField.getType(), elementType},
+            /*outputs=*/{});
+        auto pushFn = b.create<func::FuncOp>(loc, funcName, pushFnType);
+        pushFn.setPrivate();
+
+        Block *entryBlock = pushFn.addEntryBlock();
+        b.setInsertionPointToStart(entryBlock);
+        BlockArgument elementsField = pushFn.getArgument(0);
+        BlockArgument sizeField = pushFn.getArgument(1);
+        BlockArgument capacityField = pushFn.getArgument(2);
+        BlockArgument value = pushFn.getArgument(3);
+
+        Value sizeVal = b.create<memref::LoadOp>(loc, sizeField);
+        Value capacityVal = b.create<memref::LoadOp>(loc, capacityField);
+
+        Value predicate =
+            b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq, sizeVal, capacityVal);
+        b.create<scf::IfOp>(loc, predicate, [&](OpBuilder &thenBuilder, Location loc) {
+            Value two = thenBuilder.create<arith::ConstantIndexOp>(loc, 2);
+            Value newCapacity = thenBuilder.create<arith::MulIOp>(loc, capacityVal, two);
+            Value oldElements = thenBuilder.create<memref::LoadOp>(loc, elementsField);
+            Value newElements = thenBuilder.create<memref::ReallocOp>(
+                loc, cast<MemRefType>(oldElements.getType()), oldElements, newCapacity);
+            thenBuilder.create<memref::StoreOp>(loc, newElements, elementsField);
+            thenBuilder.create<memref::StoreOp>(loc, newCapacity, capacityField);
+            thenBuilder.create<scf::YieldOp>(loc);
+        });
+
+        Value elementsVal = b.create<memref::LoadOp>(loc, elementsField);
+        b.create<memref::StoreOp>(loc, value, elementsVal,
+                                  /*indices=*/sizeVal);
+
+        Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+        Value newSize = b.create<arith::AddIOp>(loc, sizeVal, one);
+
+        b.create<memref::StoreOp>(loc, newSize, sizeField);
+        b.create<func::ReturnOp>(loc);
+        return SymbolRefAttr::get(ctx, funcName);
+    }
+
+    void emitPush(Location loc, Value value, OpBuilder &b, FlatSymbolRefAttr pushFn) const
+    {
+        b.create<func::CallOp>(loc, pushFn, /*results=*/TypeRange{},
+                               /*operands=*/ValueRange{dataField, sizeField, capacityField, value});
+    }
+};
+
+struct LowerInitVector : public OpConversionPattern<VectorInitOp> {
+    using OpConversionPattern<VectorInitOp>::OpConversionPattern;
+
+    LogicalResult matchAndRewrite(VectorInitOp op, OpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const override
+    {
+        SmallVector<Type> resultTypes;
+        if (failed(getTypeConverter()->convertType(op.getType(), resultTypes))) {
+            op.emitError() << "Failed to convert type " << op.getType();
+            return failure();
+        }
+        Value capacity = rewriter.create<arith::ConstantIndexOp>(op.getLoc(), 1);
```

Comment:
We probably want to use a larger starting value for the capacity. At the very least 8, or maybe 32? What do you think?

Reply:
Sure thing, the only downside of starting larger is that we'd potentially pay for extra memory that we don't need, but 32 seems sufficiently small that it should be fine.